### ISSUE TITLE
Feed components are being re-rendered on typing

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#2185]
This issue is created as a part of this one https://github.com/appirio-tech/connect-app/issues/2138

Typing is very slow when there are several topic threads. Every time we type a symbol all the threads are rerendered. In the next video, I quickly type the same latter with the same interval but letters appear in batches rather one by one http://take.ms/wA1u2

See for example this project https://connect.topcoder-dev.com/projects/2871 and try to type a new topic title.

Aha! Link: https://topcoder.aha.io/features/TCCONNECT-418

**Comment by vikasrohit** [comment]
@maxceem are you working on it? If you haven't started it yet, please wait on this until we close the 2.4.4 milestone.

**Comment by maxceem** [comment]
@vikasrohit it's on my list, but I'm not working on it as this one is quite heavy, will most likely require to refactor the Feed and/or FeedContainer components.

**Comment by vikasrohit** [comment]
Okay, makes sense. Already moved it to 2.4.5.

**Comment by vikasrohit** [comment]
@maxceem Moving it one more release further.

**Comment by maxceem** [comment]
There are few things to keep in mind:
- We don't wanna use `shouldComponentUpdate` for fixing this as it would be hard to support it in the future. Re-renders has to be fixed in a natural way. Most likely by refactoring `FeedContainer.js` to take care about only one topic rather then the list of topics as done now. As a reference we can use `PhaseFeedHOC.jsx` which implements feeds for pahses (maybe we don't need HOC in this case, just container).
- Make sure that all existent functionality works, in particular:
    - When user started editing post and try to leave the page, confirmation prompt is appearing https://github.com/appirio-tech/connect-app/blob/dev/src/projects/detail/containers/FeedContainer.js#L423.

**Comment by sumitdaga** [comment]
please assign this to me if it is open for pick up!

**Comment by sumitdaga** [comment]
@maxceem  
ok so, so far from my analysis so far what i think may be a good idea is to not propagate "newTitle" all the way down from FeedContainer to RichTextEditor as the RichTextEditor is anyway maintaining its own title state ...this i think will avoid all re-renderings happening due to props change when only the title is being edited and not yet saved?

does it make sense?
may to have to dig a little more deeper and actually implement to see if it works

**Comment by maxceem** [comment]
@sumitdaga Yeah, I was considering before the way you proposed `not propagate "newTitle"` but it would break the work of the prompt which has to be shown when we are trying to leave the page where we started editing some post:
```
       <Prompt
          when={!!onLeaveMessage}
          message={onLeaveMessage}
        />
```
The reason for this is that we have to compare edited post with the initial post to determine if the user edited it. If we don't propagate `newTitle` we wouldn't know at the level of `FeedContainer.js` if it was changed or not.

> may to have to dig a little more deeper and actually implement to see if it works

The only appreach I can see here is instead of using `FeedContainer.js` to display the list of topics. `FeedContainer.js` has to disaplay only one topic and take care about everything for that topics. So it has to do the same what `FeedContainer.js` does but only for one topic. And after we just display the list of of topics with some new component like `DashboardPostsContainer.js` - note, that `newTitle` should not be propogating until `DashboardPostsContainer.js` thus it wouldn't force other posts to be rerendered. Only current post which is edited would be rerendred in such case.

There still could be some nuances, but this way look like have to work in general.

**Comment by sumitdaga** [comment]
@maxceem  
i understand the concern for showing   

       <Prompt  
          when={!!onLeaveMessage}  
          message={onLeaveMessage}  
        />

i was thinking of storing newTitle in the `state` of FeedContainer but not passing it down, so that it can be compared, but i guess it will cause a similar issue if have to call `mapFeed` again  

i hope i will be given more time to fix this issue

**Comment by maxceem** [comment]
> i hope i will be given more time to fix this issue
@dmessing no problem! As you updated with the valid reason. I understand that it may take some time to come up with a good solution here, so take your time.

**Comment by sumitdaga** [comment]
@maxceem  an update
i won't be able to work on this issue for about the next 24 hrs  
i leave it upto you if you want to open it up for pickup or if some delay is fine

**Comment by KevinLeigh** [comment]
@maxceem Please assign ticket to me

**Comment by maxceem** [comment]
@KevinLeigh it yours since you've commented. To make it more clear I can assign you to this issue in the interface. To do so, please, accept the invitation to this repo which has to pop up to your email. Thanks.

**Comment by KevinLeigh** [comment]
@maxceem Thank you, much appreciated. Will accept link as soon as I receive it.

**Comment by KevinLeigh** [comment]
@maxceem I have accepted the invitation.

**Comment by KevinLeigh** [comment]
@maxceem Hey there, is there a test account that I can use to test issue ?

**Comment by maxceem** [comment]
@KevinLeigh `pshah_manger`/`topcoder123`

**Comment by KevinLeigh** [comment]
@maxceem Thank you!

**Comment by KevinLeigh** [comment]
Hey @maxceem, please see PR here -> https://github.com/appirio-tech/connect-app/pull/2504, sorry it came in a bit late, only got a chance to work on it recently.

**Comment by sumitdaga** [comment]
@KevinLeigh  
refer https://github.com/appirio-tech/connect-app/issues/2185#issuecomment-418233790

@maxceem  just FYI i am free to take it if at all

**Comment by maxceem** [comment]
@KevinLeigh please, have a look on this comment in the issue https://github.com/appirio-tech/connect-app/issues/2185#issuecomment-418233790. Provided PR doesn't meet the requirement so I've closed it.

Please, let me know if you'd like to continue working on this issue and provide the PR which takes into account the requirements in the comment mentioned above.

**Comment by maxceem** [comment]
> @maxceem just FYI i am free to take it if at all

@sumitdaga thanks. It woul be good if can continue on this just in case. But I'll have to open it for everyone if **KevinLeigh** would be not interested to finish it to follow the Bug Bash rules and keep it fair for everyone.

**Comment by KevinLeighCrain** [comment]
Hey All, I am still interested in tackling this issue. I apologize for the late reply, it was after 3 in the morning here and needed some rest.

**Comment by KevinLeigh** [comment]
@maxceem Hey there, just pushed up changes, I am still using `shouldComponentUpdate` but a much cleaner way which is intended to scale. Let me know what you think.

**Comment by maxceem** [comment]
@KevinLeigh I've reviewed your PR, see https://github.com/appirio-tech/connect-app/pull/2508#pullrequestreview-154903549

It's cool to try to find an unordinary approach. But we would really like to avoid using `shouldComponentUpdate` as in the future while adding new features it would be harder to support it.  While I think it's possible to get it done faster using `shouldComponentUpdate` here we would like to refactor the component instead so it works naturally. That's why the budget for this task is relatively high.

Let me know if you still would like to try to accomplish this task, but this time strictly without `shouldComponentUpdate`.

**Comment by vigneshTheDev** [comment]
Hi, as there is no activity for 12 hours, is it open for pickup? if so, I would like to work on it

**Comment by maxceem** [comment]
As @KevinLeigh working on it, I cannot apply the 12h rule here. Because I gave him feedback at an arbitrary time which he cannot control.

If there wouldn't be any update in 24 hours since the feedback, the issue will be opened for pick up for everyone. 

@vigneshTheDev just want to mention that I cannot assign you to the issue before the current participant released issue by himself, or issue was marked as open for pick up to make it fair for all participant, see the discussion here https://github.com/appirio-tech/connect-app/issues/2279#issuecomment-419857847. 
Thanks.

**Comment by maxceem** [comment]
Due to inactivity, this issue is now open for pick up.

**Comment by arun1595** [comment]
@maxceem I would like to take a stab at this issue.

**Comment by maxceem** [comment]
Sure @arun1595.

**Comment by arun1595** [comment]
@maxceem I didn't have a chance to spend time on this task, caught up with something else. Feel free to open it up for pickup so someone else fixes the issue.

**Comment by sumitdaga** [comment]
I will again like to take a look

**Comment by maxceem** [comment]
@sumitdaga assigned it to you.

**Comment by sumitdaga** [comment]
@maxceem  
we can take two approaches ..
a) one is what you suggest, move some functions up the tree and create a new node 
b) i think we can move some functions down the tree as well and achieve what we want , because we already have a Feed component in `src/components/Feed/Feed.jsx` which is exactly/(or quite close) to what you want FeedContainer.js to become in the first approach ...  
at least i would like if we can discuss it on slack ? let me know!

**Comment by sumitdaga** [comment]
@maxceem  
as we discussed on slack ( and also came on conclusion that this cannot be achieved without `shouldComponentUpdate` or making prompt global ) ...i have raised a PR along the lines of my suggestion ... PR is at #2523 

note that its still work in progress and i just wanted to more or less show you how i think i will go about it .. so, for now i have applied fix only when you try to edit post title ....
let know what you think!

thanks

**Comment by maxceem** [comment]
@sumitdaga I've checked it https://github.com/appirio-tech/connect-app/pull/2523#pullrequestreview-156194066

**Comment by sumitdaga** [comment]
PR #2523 is finally ready for review!

**Comment by maxceem** [comment]
There are many things appears to be tights together. And it could be that direct refactoring would be too wast and thus prone to errors.
So I guess we will try to approach it from the other side and will try to move out things one by one by smaller portions keeping the functionality.
And maybe only at last refactor `FeedContainer` to remove re-renders.

The PR was closed without merging and could be used for reference https://github.com/appirio-tech/connect-app/pull/2523

Meanwhile I think we will start with solving another issue first https://github.com/appirio-tech/connect-app/issues/2556

**Comment by maxceem** [comment]
@vikasrohit found a quick fix with `shouldComponentUpdate`. But still the solution is quite clear and feels fast enough.

See the overall result demo https://monosnap.com/file/eAOgk4HfRjPqmFW9VI3PJWAwieave6

See comparison before and after of two fixes applied.

1. Before [1st fix](https://github.com/appirio-tech/connect-app/commit/4731ce2308f599aa148e82f8d7081dc39307974d), every topic is fully re-rendered, see demo video https://monosnap.com/file/kpnPOImF8BUMPAJlfXj6NHUuurdAr6
    After [1st fix](https://github.com/appirio-tech/connect-app/commit/4731ce2308f599aa148e82f8d7081dc39307974d), only the current topic and the whole container are re-rendered which works pretty fast https://monosnap.com/file/sjPNZ6GpuP3XQ0fJ79RJBGjkxMlcSv

2. Also, I've noticed that on every click on the page we also re-render all the topics, see demo videohttps://monosnap.com/file/41ihbGsA9Ph0p5gEtBhqxn9J8OOQ0Z
    After [2nd fix](https://github.com/appirio-tech/connect-app/commit/e1acdf05d0b11bd5949d2945a7b6acb2e219fb7f) it only re-renders the topic which we focused/unfocused and only when the focused status changed, see demo https://monosnap.com/file/O1vUA5H3rxIUwJPFAQXOeNGg2eRMjl.

Both fixes are quite safe to apply as hotfix as they don't have complicated logic changes.

**Comment by vikasrohit** [comment]
@maxceem I can see significant difference in dev after this patch. You can also test it here https://connect.topcoder-dev.com/projects/4828. I check this project before the patch and it was having the lag between keystroke and the actual character appearing in the text box. Let me discuss with team if we are ready for the hotfix for the production as well.

**Comment by vikasrohit** [comment]
@maxceem Lets create PR for the hotfix on master. We will merge it if we need to, otherwise we will close it when we release it with our standard release.

**Comment by maxceem** [comment]
Created PR for PROD https://github.com/appirio-tech/connect-app/pull/3006

**Comment by vikasrohit** [comment]
Thanks @maxceem

**Comment by vikasrohit** [comment]
@maxceem do you think we still have opportunity to improve it? I am asking from perspective of closing this github issue.

**Comment by maxceem** [comment]
@vikasrohit theoretically there is a way to improve it and avoid using `shouldComponentUpdate` with comparing properties. But most likely such improvement will require a wast refactoring.
As per me it feels that the current speed is quite comfortable and I don't feel an issue with typing speed.

So I think we can close it for now and create a new issue later if there is any feedback that typing experience is not smooth enough.

**Comment by vikasrohit** [comment]
Closing as solution is merged into dev already and would move to the production with our regular release cycle.
